File: app/activities/openai_activities.py
# AIDEV-NOTE: Generic OpenAI activities for Temporal workflows
import os
import base64
import logging
from openai import AsyncOpenAI
from temporalio import activity
import mimetypes

from app.models.ai_models import AIRequest, AIResponse, OpenAIActivities

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class OpenAIActivitiesImpl(OpenAIActivities):
    """Implementation of OpenAI activities"""
    
    def __init__(self):
        # Configure OpenAI Client
        api_key = os.getenv("OPENAI_API_KEY")
        if not api_key:
            raise ValueError("OPENAI_API_KEY environment variable not set.")
        
        self.client = AsyncOpenAI(api_key=api_key)
        self.default_model = "gpt-4-turbo-preview"
        logger.info("OpenAI Activities initialized")

    @activity.defn
    async def process_request(self, request: AIRequest) -> AIResponse:
        """Process a request using OpenAI"""
        activity.logger.info(f"Processing OpenAI request with prompt: {request.prompt[:100]}...")
        
        try:
            messages = []
            
            # Handle file input if provided
            if request.file_path:
                with open(request.file_path, "rb") as f:
                    file_bytes = f.read()
                
                mime_type, _ = mimetypes.guess_type(request.file_path)
                
                # AIDEV-NOTE: OpenAI supports image files for vision models
                if mime_type and mime_type.startswith('image/'):
                    base64_image = base64.b64encode(file_bytes).decode('utf-8')
                    messages.append({
                        "role": "user",
                        "content": [
                            {"type": "text", "text": request.prompt},
                            {
                                "type": "image_url",
                                "image_url": {
                                    "url": f"data:{mime_type};base64,{base64_image}"
                                }
                            }
                        ]
                    })
                    model = "gpt-4-vision-preview"  # Use vision model for images
                else:
                    # For non-image files, include content as text if possible
                    try:
                        file_content = file_bytes.decode('utf-8')
                        messages.append({
                            "role": "user",
                            "content": f"{request.prompt}\n\nFile content:\n{file_content}"
                        })
                    except:
                        messages.append({
                            "role": "user",
                            "content": f"{request.prompt}\n\n[Binary file provided: {request.file_path}]"
                        })
                    model = self.default_model
            else:
                messages.append({"role": "user", "content": request.prompt})
                model = self.default_model
            
            # Apply custom parameters
            params = {
                "model": model,
                "messages": messages,
                "temperature": 0.7,
                "max_tokens": 4096,
            }
            
            if request.parameters:
                params.update(request.parameters)
                if "model" in request.parameters:
                    model = request.parameters["model"]
            
            # Make API call
            response = await self.client.chat.completions.create(**params)
            
            return AIResponse(
                content=response.choices[0].message.content,
                model_used=model,
                tokens_used=response.usage.total_tokens if response.usage else None,
                metadata={
                    "finish_reason": response.choices[0].finish_reason,
                    "prompt_tokens": response.usage.prompt_tokens if response.usage else None,
                    "completion_tokens": response.usage.completion_tokens if response.usage else None,
                }
            )
            
        except Exception as e:
            activity.logger.error(f"Error processing OpenAI request: {e}", exc_info=True)
            raise

File: src/TemporalAI/Activities/OpenAIActivities.cs
// AIDEV-NOTE: Generic OpenAI activities for Temporal workflows
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using OpenAI;
using OpenAI.Chat;
using OpenAI.Images;
using TemporalAI.Models;
using Temporalio.Activities;

namespace TemporalAI.Activities
{
    /// <summary>
    /// Implementation of OpenAI activities
    /// </summary>
    public class OpenAIActivities : IOpenAIActivities
    {
        private readonly ILogger<OpenAIActivities> _logger;
        private readonly OpenAIClient _client;
        private readonly string _defaultModel = "gpt-4-turbo-preview";

        public OpenAIActivities(ILogger<OpenAIActivities> logger)
        {
            _logger = logger;
            
            var apiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");
            if (string.IsNullOrEmpty(apiKey))
            {
                throw new InvalidOperationException("OPENAI_API_KEY environment variable not set.");
            }

            _client = new OpenAIClient(apiKey);
            _logger.LogInformation("OpenAI Activities initialized");
        }

        [Activity]
        public async Task<AIResponse> ProcessRequestAsync(AIRequest request)
        {
            var context = ActivityExecutionContext.Current;
            _logger.LogInformation("Processing OpenAI request with prompt: {Prompt}", 
                request.Prompt.Length > 100 ? request.Prompt.Substring(0, 100) + "..." : request.Prompt);

            try
            {
                var messages = new List<ChatMessage>();
                var model = _defaultModel;

                // Handle file input if provided
                if (!string.IsNullOrEmpty(request.FilePath))
                {
                    var fileBytes = await File.ReadAllBytesAsync(request.FilePath);
                    var mimeType = GetMimeType(request.FilePath);

                    // AIDEV-NOTE: OpenAI supports image files for vision models
                    if (mimeType.StartsWith("image/"))
                    {
                        var base64Image = Convert.ToBase64String(fileBytes);
                        var imageUrl = $"data:{mimeType};base64,{base64Image}";
                        
                        messages.Add(new UserChatMessage(
                            new List<ChatMessageContentPart>
                            {
                                ChatMessageContentPart.CreateTextMessageContentPart(request.Prompt),
                                ChatMessageContentPart.CreateImageMessageContentPart(
                                    new Uri(imageUrl),
                                    ImageChatMessageContentPartDetail.Auto
                                )
                            }
                        ));
                        model = "gpt-4-vision-preview"; // Use vision model for images
                    }
                    else
                    {
                        // For non-image files, include content as text if possible
                        try
                        {
                            var fileContent = System.Text.Encoding.UTF8.GetString(fileBytes);
                            messages.Add(new UserChatMessage($"{request.Prompt}\n\nFile content:\n{fileContent}"));
                        }
                        catch
                        {
                            messages.Add(new UserChatMessage($"{request.Prompt}\n\n[Binary file provided: {request.FilePath}]"));
                        }
                    }
                }
                else
                {
                    messages.Add(new UserChatMessage(request.Prompt));
                }

                // Apply custom parameters
                var temperature = 0.7f;
                var maxTokens = 4096;

                if (request.Parameters != null)
                {
                    if (request.Parameters.TryGetValue("temperature", out var tempValue) && tempValue is double temp)
                        temperature = (float)temp;
                    if (request.Parameters.TryGetValue("max_tokens", out var tokensValue) && tokensValue is int tokens)
                        maxTokens = tokens;
                    if (request.Parameters.TryGetValue("model", out var modelValue) && modelValue is string modelName)
                        model = modelName;
                }

                // Create chat options
                var options = new ChatCompletionOptions
                {
                    Temperature = temperature,
                    MaxTokens = maxTokens
                };

                // Make API call
                var chatClient = _client.GetChatClient(model);
                var response = await chatClient.CompleteChatAsync(messages, options);
                var completion = response.Value;

                return new AIResponse
                {
                    Content = completion.Content[0].Text,
                    ModelUsed = model,
                    TokensUsed = completion.Usage?.TotalTokens,
                    Metadata = new Dictionary<string, object>
                    {
                        ["finish_reason"] = completion.FinishReason.ToString(),
                        ["prompt_tokens"] = completion.Usage?.InputTokens ?? 0,
                        ["completion_tokens"] = completion.Usage?.OutputTokens ?? 0
                    }
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error processing OpenAI request");
                throw;
            }
        }

        private static string GetMimeType(string filePath)
        {
            var extension = Path.GetExtension(filePath).ToLowerInvariant();
            return extension switch
            {
                ".jpg" or ".jpeg" => "image/jpeg",
                ".png" => "image/png",
                ".gif" => "image/gif",
                ".webp" => "image/webp",
                ".pdf" => "application/pdf",
                ".txt" => "text/plain",
                ".json" => "application/json",
                ".xml" => "application/xml",
                _ => "application/octet-stream"
            };
        }
    }
}